internal/nvmeof/subsystem_tracker.go

```diff
@@ -0,0 +1,308 @@
+
+		return true, 0, nil // Consider it as "should disconnect"
+	}
+
+	// Decrement reference count
+	if subsysState.ReferenceCount > 0 {
+		subsysState.ReferenceCount--
+	}
+	removeVolumeFromSubsystem(subsysState, volumeID)
+	timeNow := time.Now()
+	subsysState.LastUpdated = timeNow
+	t.state.LastUpdate = timeNow
+
+	shouldDisconnect := subsysState.ReferenceCount == 0
+
+	// Remove from state if no more references
+	if shouldDisconnect {
+		delete(t.state.Subsystems, subsystemNQN)
+	}
+
+	// Save state
+	if err := t.saveStateLocked(); err != nil {
+		// Rollback on failure
+		if shouldDisconnect {
+			t.state.Subsystems[subsystemNQN] = subsysState
+		}
+		subsysState.ReferenceCount++
+		addVolumeToSubsystem(subsysState, volumeID) // rollback
+
+		return false, subsysState.ReferenceCount, fmt.Errorf("failed to save state after removing reference: %w", err)
+	}
+
+	log.DebugLog(ctx, "Removed reference for subsystem %s, , volume %s, remaining count: %d, should disconnect: %v",
+		subsystemNQN, volumeID, subsysState.ReferenceCount, shouldDisconnect)
+
+	return shouldDisconnect, subsysState.ReferenceCount, nil
+}
+
+// addVolumeToSubsystem associates a volume ID with a subsystem for tracking.
+func addVolumeToSubsystem(subsysState *subsystemState, volumeID string) {
+	// Add volume if not already present
+	if slices.Contains(subsysState.Volumes, volumeID) {
+		return // Already present
+	}
+	subsysState.Volumes = append(subsysState.Volumes, volumeID)
+}
+
+// removeVolumeFromSubsystem removes a volume ID from a subsystem.
+func removeVolumeFromSubsystem(subsysState *subsystemState, volumeID string) {
+	// Remove volume from list
+	newVolumes := make([]string, 0, len(subsysState.Volumes))
+	for _, vol := range subsysState.Volumes {
+		if vol != volumeID {
+			newVolumes = append(newVolumes, vol)
+		}
+	}
+	subsysState.Volumes = newVolumes
+}
+
+// GetState returns a copy of the current tracker state for debugging.
+func (t *fileBasedTracker) GetState(ctx context.Context) map[string]interface{} {
+	t.mutex.RLock()
+	defer t.mutex.RUnlock()
+
+	state := make(map[string]interface{})
+	state["version"] = t.state.Version
+	state["node_id"] = t.state.NodeID
+	state["last_update"] = t.state.LastUpdate
+	state["subsystem_count"] = len(t.state.Subsystems)
+
+	subsystems := make(map[string]interface{})
+	for nqn, subsys := range t.state.Subsystems {
+		subsystems[nqn] = map[string]interface{}{
+			"reference_count": subsys.ReferenceCount,
+			"last_updated":    subsys.LastUpdated,
+			"volume_count":    len(subsys.Volumes),
+		}
+	}
+	state["subsystems"] = subsystems
+
+	return state
+}
+
+// loadState loads the subsystem state from disk.
+func (t *fileBasedTracker) loadState() error {
+	// Check if state file exists
+	if _, err := os.Stat(t.stateFile); os.IsNotExist(err) {
+		log.DefaultLog("State file %s does not exist, starting with empty state", t.stateFile)
+
+		return nil
+	}
+
+	// Read state file
```

Comment:
While reading and writing the file, there should be a file-lock used. It is unlikely, but technically possible that two node-plugins run on the same system (possibly with different drivernames). When one process is writing, an other reading process may read incomplete/incorrect data.

Maybe instead of the Mutex in the `fileBasedTracker`, a file lock can be used?